pkg/controller/networkpolicy/validate.go

```diff
@@ -760,6 +766,29 @@ func (v *antreaPolicyValidator) validateMulticastIGMP(ingressRules, egressRules
 	return "", true
 }
 
+// validateL7Protocols validates the action field is only set as Allow when L7Protocols
+// field is set in Antrea-native policy rules.
+func (v *antreaPolicyValidator) validateL7Protocols(ingressRules, egressRules []crdv1alpha1.Rule) (string, bool) {
+	for _, r := range append(ingressRules, egressRules...) {
+		if len(r.L7Protocols) > 0 {
+			if *r.Action != crdv1alpha1.RuleActionAllow {
+				return "layer 7 protocols can only be used with Allow action", false
+			}
+			for _, port := range r.Ports {
+				if port.Protocol != nil && *port.Protocol != v1.ProtocolTCP {
+					return "layer 7 protocols can only be used when layer 4 protocol is TCP or unset", false
```

Comment:
not all layer 7 protocols are based on TCP, though the only supported protocol http is. To avoid confusion and to be extensible, it should validate the l4 protocol is either unset or TCP if the l7 protocol is http, and the error message should say "HTTP protocol can only be used when layer 4 protocol is TCP or unset"

Reply:
Updated for protocols and ports. Added verification for toServices.